### PR TITLE
Add optax optimizer support to trainer

### PR DESCRIPTION
**Pull request by lienchen0526**
This change introduced the usage of Optax Optimizer, allowing developer to pass a `GradientTransformation` to the `Trainer`.

### Canonical Usage

```python
import optax

LEARNING_RATE = 1e-2
LOG_STEPS_INTERVAL = 5

# Create Optax optimizer
optax_optimizer = optax.chain(
    optax.adamw(LEARNING_RATE, weight_decay=.01),
)

# Initialize trainer
trainer = Trainer(
    model=model,
    optimizer=optax_optimizer,
    train_dataloader=train_dataloader,
    epochs=1,
    log_steps_interval=LOG_STEPS_INTERVAL,
    tensorboard_dir=os.path.join(self.TMP_DIR, "tensorboard")  # Note that tensorboard is not supported yet.
)

# Start training
trainer.train()
```

### Sources

* Introduced `KitharaOptimizer` as the base class of the optimizer management class internal used by Kithara.
* Separated the logic from `keras.Optimizer` if developer pass the `optax.GradientTransformation` in `Trainer`.

### Limitation

* Not yet supporting tensorboard
* Not yet supporting learning rate tracking

### Todo

* Migrate the implementation of `keras.Optimizer` to internal `KitharaOptimizer`
* Resolve limitations

### Unit Tests
* Added a test of using `optax.adam` to `Trainer` (`test_trainer_creation.py`)
* Added a test to validate the protocol integration of `KitharaOptimizer`. (`test_protocol.py`)
* Added a test suite to validate the behaviors of `OptaxOptimizerInterface`. (`test_optax_optimizer.py`)

### CODE REVIEW

**Comment by wenxindongwork** [comment]
Thanks so much Jerry!

As someone who is not super familiar with optax optimizer, I guess the following will both be supported?

```
# Chained optimizers
optax_optimizer = optax.chain(
    optax.adamw(LEARNING_RATE, weight_decay=.01),
    optax.scale_by_schedule(decay_schedule)  # or some other transformations
)

# Single optimzer
optax_optimizer = optax.adamw(LEARNING_RATE, weight_decay=.01)
```

File: kithara/dataset/text_completion.py
Comment (wenxindongwork) [inline kithara/dataset/text_completion.py:1]: Could you update the trainer api here? 

https://github.com/AI-Hypercomputer/kithara/blob/main/docs/source/api/kithara.trainer_api.rst

Though we really should set up auto code -> doc at some point...

File: kithara/optimizers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from jaxtyping import PyTree
+import keras
+from kithara.optimizers.protocol import KitharaOptimizer
+from kithara.optimizers.optax_optimizer import OptaxOptimizerInterface
+import optax
+
+def convert_to_kithara_optimizer(
```
Comment (wenxindongwork) [inline kithara/optimizers/__init__.py:6]: If this function is a converter for only the optax optimizer, could we remove `keras.Optimizer` from the typing?

File: kithara/trainer/trainer.py [hunk 2]
```diff
@@ -44,7 +47,7 @@ class Trainer:
 
     Attributes:
         model (kithara.Model): The model to be trained or evaluated.
-        optimizer (keras.Optimizer): The optimizer used for training.
+        optimizer (keras.Optimizer | OptaxOptimizerConfig): The optimizer used for training.
```
Comment (wenxindongwork) [inline kithara/trainer/trainer.py]: I notice that the typing here is OptaxOptimizerConfig, and below it is optax.GradientTransformation, I wonder how do they differ

File: kithara/trainer/trainer.py [hunk 3]
```diff
@@ -203,9 +211,13 @@ def _train_step(self, state: Tuple[List[jax.Array]], data: dict):
...
-        trainable_variables, optimizer_variables = self.optimizer.stateless_apply(
-            optimizer_variables, grads, trainable_variables
-        )
+        if isinstance(self.optimizer, KitharaOptimizer):
+            trainable_variables, optimizer_variables = self.optimizer.stateless_apply(
+                trainable_variables, grads, optimizer_variables)
+        else:
+            trainable_variables, optimizer_variables = self.optimizer.stateless_apply(
+                optimizer_variables, grads, trainable_variables
+            )
```
Comment (wenxindongwork) [inline kithara/trainer/trainer.py]: I might be overlooking something here, but wondering why can't we call 
```
self.optimizer.stateless_apply(
                optimizer_variables, grads, trainable_variables
            )
``` 

for all cases since at this point all optimizers should have been converted to a KitharaOptimizer.

File: kithara/trainer/trainer.py [hunk 4]
```diff
@@ -294,7 +306,8 @@ def train(self):
                     "samples_per_second": round(samples_per_second, 2),
                     "train_steps_per_second": round(1 / step_time, 2),
                     "samples_seen": self.global_batch_size * self.step_count,
-                    "learning_rate": self.optimizer.learning_rate.value,
+                    "learning_rate": self.optimizer.learning_rate.value if isinstance(
```
Comment (wenxindongwork) [inline kithara/trainer/trainer.py]: in the PR description you mentioned that tensorboard and learning rate tracking are not currently supported, but here we are logging the learning rate right? so, is it only tensorboard that is not currently being supported?

File: kithara/optimizers/keras_optimizer.py
Comment (wenxindongwork) [inline kithara/optimizers/keras_optimizer.py:1]: should we delete this file?

File: kithara/optimizers/optax_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    def learning_rate(self) -> None:
+        return
+
+    @property
+    def state_or_variables(self):
+        return self.state
+
+    @state_or_variables.setter
+    def state_or_variables(self, state: PyTree):
+        self.state = state
```
Comment (wenxindongwork) [inline kithara/optimizers/optax_optimizer.py]: I'm a bit worried right now that the optimizer variables are not being sharded. 

One solutions is to convert each jax array in this state into a Keras Variable. 

You can check out this example: 
https://keras.io/api/layers/backend_specific_layers/flax_layer/
https://github.com/keras-team/keras/blob/master/keras/src/layers/layer.py#L556-L565

This way, we have 
- a unified way to shard optimizer variables using Keras global state
- a unified way to verify that the optimizer variables are sharded correctly, and calculate optimizer size
- can get rid of a lot of if-else conditions in the trainer loop.
- get tensorboard support for free. 

We should expect all optimizers to convert their state into a pytree of Keras Variables.

File: kithara/dataset/text_completion.py
Comment (lienchen0526) [reply kithara/dataset/text_completion.py:1]: resolved in the next commit

File: kithara/optimizers/__init__.py (hunk 1 above)
Comment (lienchen0526) [reply kithara/optimizers/__init__.py:6]: Yes, we could. And I believe we can convert the optimizer to make it has similar interface with `keras.optimizers.Optimzer`. The detail implementation is in next commit.

File: kithara/optimizers/keras_optimizer.py
Comment (lienchen0526) [reply kithara/optimizers/keras_optimizer.py:1]: Yes, we can. I believe it can be more clear. I delete it in the next commit.

File: kithara/optimizers/optax_optimizer.py (hunk 5 above)
Comment (lienchen0526) [reply kithara/optimizers/optax_optimizer.py]: I believe it is a good approach to convert the optax state to `keras.Variable`. In the next commit, I implemented this features and make the optax state transformed into `Pytree[keras.Variable]`. And I made the variable being distributed just like the trainable variables.

But the following limitations still exsist:
* tensorboard is limited since the learning rate of optax optimizer is not tracked by default.

File: kithara/trainer/trainer.py (hunk 2 above)
Comment (lienchen0526) [reply kithara/trainer/trainer.py]: done. Thank you WenXin!

File: kithara/trainer/trainer.py (hunk 3 above)
Comment (lienchen0526) [reply kithara/trainer/trainer.py]: yes, this is quite a good point. I updated the change in the next commit.

File: kithara/trainer/trainer.py (hunk 4 above)
Comment (lienchen0526) [reply kithara/trainer/trainer.py]: It is still not supported since the `learning_rate` is `None`. This will cause exceptions when using the tensorboard callback because the tensorboard expected values in `learning_rate`.

**Review by lienchen0526 — COMMENTED** [review]
The comments are mostly resolved. The suggested implementation is in the next commit: https://github.com/AI-Hypercomputer/kithara/pull/8/commits/c2ae0ae0c5bfb142f47f1b496d4f4aafc5d911f3

File: kithara/optimizers/optax_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        def create_keras_variable_pytree(
+                subtree, trainable_vars) -> PyTree[backend.Variable]:
+
+            if not is_same_shape_with_trainable_subtree(subtree):
+                return jax.tree.map(
+                    lambda leaf: backend.Variable(leaf,), subtree
+                )
+
+            return jax.tree.map(
+                lambda leaf, keras_var: backend.Variable(
```
Comment (wenxindongwork) [inline kithara/optimizers/optax_optimizer.py:90]: could we add a comment here to give an example of what the keras variable name for these optimizer variables look like? 

E.g. decoder-block-0/layer-1/kernel -> decoder-block-0_layer-1_kernel_optstate

File: kithara/optimizers/protocol.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+from typing import (
+    Protocol, Tuple, TypeAlias,
+    runtime_checkable)
+
+from jaxtyping import PyTree
+from keras.src.backend import Variable
+
+TrainableVariable: TypeAlias = PyTree[Variable]
+OptimizerVariable: TypeAlias = PyTree[Variable]
```
Comment (wenxindongwork) [inline kithara/optimizers/protocol.py:9]: the Keras convention is to represent model and optimizer variables as flat lists of jax arrays. I think it's fine if we use PyTree here, but if we use flat lists (and write converters for Optimizer Keras flat list <> Jax Pytree) we can simplify the syntax in the main training loop and make it look a little bit less scary and more hackable for people who don't use JAX. It could also preserve the consistency between model variables and optimizer variables being the same format. 

The converters could look something like this (MaxTextConversionMixIn.get_maxtext_params) just for an idea: 

```
        """Convert Kithara variables (flat list format) into 
        MaxText variables (nested dict format). This function 
        is a simple format converter. It is used for inserting 
        the current model parameters into MaxText's inference 
        engine. 
        """
        params = {}
        for v in model.variables:
            variable_name = v.path
            nest_keys = variable_name.split("/")[-1].split("-")
            if nest_keys[0] != "params":
                continue
                
            current = params
            for key in nest_keys[:-1]:  # All keys except the last one
                current = current.setdefault(key, {})
            current[nest_keys[-1]] = v.value
            
        return params
 ```

let me know what you think!

File: kithara/optimizers/optax_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+from kithara.optimizers.protocol import KitharaOptimizer, TrainableVariable, OptimizerVariable
+
+
+class OptaxOptimizerInterface(KitharaOptimizer):
+    def __init__(self, optimizer: optax.GradientTransformation, params: PyTree[backend.Variable]):
+        self.optimizer, state = optimizer, optimizer.init(params)
+        self.state_in_keras_variable = self._convert_state_to_keras_variables(state, params)
+
+        with backend.name_scope("OptaxOptimizerInterface"):
```
Comment (wenxindongwork) [inline kithara/optimizers/optax_optimizer.py:16]: Didn't know this was a thing, this is cool!

**Review by wenxindongwork — COMMENTED** [review]
Thank you Jerry!! Everything look great to me, left some small comments.

File: kithara/optimizers/protocol.py (hunk 7 above)
Comment (lienchen0526) [reply kithara/optimizers/protocol.py:9]: Thanks for the info. If so, I think it is good to make `TrainableVariable` to `List[Variable]`. But for the Optimizer variables, I'll not suggest to do this since the optax state is a nested `PyTree` according to the `optax.chain`, which the shape is determined by user in the runtime. For example, if we are using:

```python
import jax
import optax

# Suppose the structure of trainable_variables is PyTreeDef([*, *, *])
trainable_variables: List[keras.Variable] = model.trainable_variables

opt = optax.chain(
  optax.adamw(.01),
  optax.clip(.1)
)

state = opt.init(trainable_variables)
_, state_treedef = jax.tree_util.tree_flatten(state)
print(treedef)  #PyTreeDef([* [*, *, *,], [*, *, *]])
```

So I will still recommend to make it a `PyTree` for optimizer variable here. This is also the reason I use `jax.tree.map` for optimizer in the `trainer.py`.

File: kithara/optimizers/optax_optimizer.py (hunk 8 above)
Comment (lienchen0526) [reply kithara/optimizers/optax_optimizer.py:16]: cheers!

File: kithara/optimizers/optax_optimizer.py (hunk 6 above)
Comment (lienchen0526) [reply kithara/optimizers/optax_optimizer.py:90]: Yes, of course. But I prefer to create the docstring at the head of this method. will do it in the next commit.

**Review by lienchen0526 — COMMENTED** [review]
I added the docstring to this method `_convert_state_to_keras_variables` to explain the behaviors, but also update the type of trainable variables to make it more straightforward.

However, for the concerning of using `jax.tree.map`, I think it may be the necessary evil due to the nature of `optax`.

File: kithara/optimizers/protocol.py (hunk 7 above)
Comment (wenxindongwork) [reply kithara/optimizers/protocol.py:9]: Got it, this makes a lot of sense, thank you for explaining!

File: kithara/optimizers/optax_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        updates, opt_state = self.optimizer.update(
+            gradients, optimizer_variable, trainable_variables)
+        new_trainable_variables = optax.apply_updates(trainable_variables, updates)
+
+        return new_trainable_variables, opt_state
+
+    def get_optimizer_memory_usage(self):
+        return sum([leaf.nbytes for leaf in jax.tree_util.tree_leaves(self.variables)])
+
+    def _convert_state_to_keras_variables(
```
Comment (wenxindongwork) [inline kithara/optimizers/optax_optimizer.py:46]: thanks for adding the detailed comment!

**Review by wenxindongwork — APPROVED** [review]
